### PR TITLE
fix(split-storage) - optimize copying of state

### PR DESCRIPTION
**Pull request by wacban**
The current approach means that when looking for a trie node in the db we need to check all shards. In practice most nodes are only ever present in one shard so the current implementation is ~4x slower when it comes to reading trie nodes than it could be. The state reads also happen to account for majority of time spent in copying heights from hot to cold. 

This fix should resolve the ever growing cold head lag that is described in more detail on [zulip](https://near.zulipchat.com/#narrow/stream/308695-pagoda.2Fprivate/topic/Split.20storage.20nodes.20increasing.20cold.20head.20lag) which is a major issue for our archival nodes.

### CODE REVIEW

**Comment by wacban** [comment]
still work in progress, please consider it a draft for now

**Comment by wacban** [comment]
ready for review

tldr of implementation
* inlined inserting into cold db inside of the processing of TrieNodeOrValueHash
* extracted that into a dedicated method called `copy_state_from_store`

notes
* the same node can be present in more than one shard - I believe then it would also be present in the TrieChanges for all the relevant shards
* we no longer extract the keys for TrieNodeOrValueHash so no other column can use it as a part of key - it is the case today and I added a test to ensure nobody breaks it by accident in the future

**Review by posvyatokum — APPROVED** [review]
LGTM. For additional optimisation look at the comment.

File: core/store/src/cold_storage.rs [hunk 1]
```diff
@@ -109,6 +117,58 @@ fn rc_aware_set(
...
+
+        let Some(trie_changes) = trie_changes else { continue };
+        for op in trie_changes.insertions() {
+            hot_store.insert_state_to_cache_from_op(op, &shard_uid_key);
+
+            let key = join_two_keys(&shard_uid_key, op.hash().as_bytes());
+            let value = hot_store.get(DBCol::State, &key)?;
+            let value =
+                value.ok_or_else(|| io::Error::new(io::ErrorKind::NotFound, hex::encode(&key)))?;
+            rc_aware_set(&mut transaction, DBCol::State, key, value);
```
[the remark this answers is not in the document]
Comment (wacban) [reply core/store/src/cold_storage.rs]: some tests started failing after removing it and we need the fix asap, I'll skip this for now
